Meta/CMake/utils.cmake

```diff
@@ -169,13 +169,15 @@ function(embed_resource target section file)
 endfunction()
 
 function(generate_state_machine source header)
-    set(source ${CMAKE_CURRENT_SOURCE_DIR}/${source})
-    add_custom_command(
-        OUTPUT ${header}
-	COMMAND ${write_if_different} ${header} ${CMAKE_BINARY_DIR}/Userland/DevTools/StateMachineGenerator/StateMachineGenerator ${source} > ${header}
-        VERBATIM
-        DEPENDS StateMachineGenerator
-        MAIN_DEPENDENCY ${source}
-    )
-    get_filename_component(output_name ${header} NAME)
+    get_filename_component(header_name ${header} NAME)
+    set(target_name "generate_${header_name}")
+    if(NOT TARGET ${target_name})
+        set(source ${CMAKE_CURRENT_SOURCE_DIR}/${source})
+        add_custom_target(${target_name}
+            BYPRODUCTS ${header}
+            COMMAND ${write_if_different} ${CMAKE_CURRENT_SOURCE_DIR}/${header} ${CMAKE_BINARY_DIR}/Userland/DevTools/StateMachineGenerator/StateMachineGenerator ${source}
+            VERBATIM
+            DEPENDS StateMachineGenerator ${source}
+        )
+    endif()
```

Comment:
Looks like this generally does the same as `compile_ipc` and `compile_gml`, but differs a bit in the implementation (`add_custom_command()` + `add_custom_target()`) - I don't know enough CMake to judge it, but unless it's better in a way I'd prefer sticking with the existing pattern (or updating the others).

Reply:
I'm not sure what you mean?

Reply:
Oh I see, you're saying I should use `add_custom_command()` and then make a target with `add_custom_target()`? Sure.

Reply:
`compile_ipc` and `compile_gml` also invoke a code generator to generate some output, and also need to declare dependencies to not mess up build order. So in a way they do pretty much the same as this thing, but are implemented differently. Question is, can we rewrite this in the existing pattern (code below) or would that not work for some reason? Just seems confusing to have different impls of things that don't do different things, essentially.

```cmake
function(compile_state_machine source output)
    set(source ${CMAKE_CURRENT_SOURCE_DIR}/${source})
    add_custom_command(
        OUTPUT ${output}
        COMMAND ${write_if_different} ${output} ${CMAKE_BINARY_DIR}/Userland/DevTools/StateMachineGenerator/StateMachineGenerator ${source}
        VERBATIM
        DEPENDS StateMachineGenerator
        MAIN_DEPENDENCY ${source}
    )
    get_filename_component(output_name ${output} NAME)
    add_custom_target(generate_${output_name} DEPENDS ${output})
endfunction()
```

Reply:
Sure we can, the thing I had was doing the same thing, in less words.
But since there's already more of this style, let's just keep using it.